spec/std/openssl/ssl/hostname_validation_spec.cr

```diff
@@ -0,0 +1,99 @@
+require "spec"
+require "openssl/ssl/hostname_validation"
+
+def openssl_create_cert(subject = nil, san = nil)
+  cert = OpenSSL::X509::Certificate.new
+  cert.subject = subject if subject
+  cert.add_extension(OpenSSL::X509::Extension.new("subjectAltName", san)) if san
+  cert.to_unsafe
+end
+
+describe OpenSSL::SSL::HostnameValidation do
+  describe "validate_hostname" do
+    it "matches IP from certificate SAN entries" do
+      OpenSSL::SSL::HostnameValidation.validate_hostname("192.168.1.1", openssl_create_cert(san: "IP:192.168.1.1")).should eq(OpenSSL::SSL::HostnameValidation::Result::MatchFound)
+      OpenSSL::SSL::HostnameValidation.validate_hostname("192.168.1.2", openssl_create_cert(san: "IP:192.168.1.1")).should eq(OpenSSL::SSL::HostnameValidation::Result::MatchNotFound)
+      OpenSSL::SSL::HostnameValidation.validate_hostname("::1", openssl_create_cert(san: "IP:::1")).should eq(OpenSSL::SSL::HostnameValidation::Result::MatchFound)
+      OpenSSL::SSL::HostnameValidation.validate_hostname("::1", openssl_create_cert(san: "IP:::2")).should eq(OpenSSL::SSL::HostnameValidation::Result::MatchNotFound)
+      OpenSSL::SSL::HostnameValidation.validate_hostname("0:0:0:0:0:0:0:1", openssl_create_cert(san: "IP:::1")).should eq(OpenSSL::SSL::HostnameValidation::Result::MatchFound)
+      OpenSSL::SSL::HostnameValidation.validate_hostname("fe80:0:0:0:0:0:0:1", openssl_create_cert(san: "IP:fe80::1")).should eq(OpenSSL::SSL::HostnameValidation::Result::MatchFound)
+      OpenSSL::SSL::HostnameValidation.validate_hostname("fe80:0:0:0:0:0:0:2", openssl_create_cert(san: "IP:fe80::1")).should eq(OpenSSL::SSL::HostnameValidation::Result::MatchNotFound)
+      OpenSSL::SSL::HostnameValidation.validate_hostname("fe80:0:1", openssl_create_cert(san: "IP:fe80:0::1")).should eq(OpenSSL::SSL::HostnameValidation::Result::MatchNotFound)
```

Comment:
Oh wait, while this one is properly a good one too, `fe80:0:1` is not actually valid, right? I was more looking towards different non-expanded formattings of the same IP, to repeat for example `fe80:0::1` and `fe80::0:1`, which is both a representation for `fe80:0:0:0:0:0:0:1`